feat(instrumentation-openai): support Responses API

btomaj opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/guides/contributor#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

 - The existing OpenAI instrumentation covers Chat Completions and Embeddings but emits no telemetry for the newer Responses API. This PR adds spans, logs, and metrics for the Responses API.

## Short description of the changes

  - Adopt Semantic Conventions 1.38.0 for Responses API.
  - Replace hard-coded values with constants from Semantic Conventions 1.38.0.
  - Add instrumentation for `OpenAI.Responses.create`:
      1. Spans
        - Calls to `Responses.create` and `Responses.stream` (as a result of internally calling `Reponses.create`) start a client span, and records the request metadata (`model`, `max_output_tokens`, `temperature`, `top_p`, base URL), and if provided, system instructions.
        - Records the response ID, finish reason, and if provided, conversation ID.
      2. Logs
        - Emits input messages as an `gen_ai.client.inference.operation.details` event at request time.
        - For streaming responses, emits individual output messages as they arrive as an `gen_ai.client.inference.operation.details` event.
        - For non-streaming responses, emits the complete output messages on completion as an `gen_ai.client.inference.operation.details` event.
        - Tool/file-search/MCP/computer/code-interpreter/image-generation items are normalized into the [GenAI Output Message JSON schema](https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-output-messages.json).
        - Message bodies are only included when `captureMessageContent` is enabled.
      4. Metrics – streaming and non-streaming flows record two histograms:
        - `gen_ai.client.operation.duration` (one sample per call, tagged with error type on failures)
        - `gen_ai.client.token.usage` (two samples per call—input vs output tokens—tagged with the response model and token type).
  - Add tests for `OpenAI.Responses.create` instrumentation:
    - Streaming and non-streaming Responses API scenarios with and without content capture.
    - Nock mocks for text, tool, file-search, MCP, computer, code-interpreter, and image-generation scenarios.

btomaj committed:
feat(instrumentation-openai): support Responses API

btomaj committed:
fix(instrumentation-openai): fix isEnabled() check

btomaj commented:
@trentm @seemk can I please request a review for this PR?

btomaj commented:
@tedsuo how do I get this PR merged? The CONTRIBUTING.md instructs to ping [@open-telemetry/javascript-approvers](https://github.com/orgs/open-telemetry/teams/javascript-approvers), but that group doesn't exist anymore as far as I can tell.

pichlermarc reviewed (commented):
Hi @btomaj. @trentm and @seemk (assigned to this PR) own this component, they are the contacts to review the PR - additionally any member of @open-telemetry/javascript-approvers can review it as well.

Although I’m not deeply familiar with this part of the codebase, I imagine the size of the PR (12k lines added across 49 files) might be making it a bit daunting to review. To help make the process more manageable, it could be worth considering splitting it into smaller, incremental PRs - starting with a minimal viable implementation, and then layering on additional features.

hectorhdzg commented on packages/instrumentation-openai/src/semconv.ts:
```diff
@@ -21,16 +21,218 @@
  *
  * @experimental This attribute is experimental and is subject to breaking changes in minor releases of `@opentelemetry/semantic-conventions`.
+ */
+export const ATTR_GEN_AI_AGENT_DESCRIPTION = 'gen_ai.agent.description' as const;
```
Most of these are already available in @opentelemetry/semantic-conventions/incubating why adding these again here?

btomaj replied:
From https://github.com/open-telemetry/opentelemetry-js/tree/main/semantic-conventions#unstable-semconv
> Because the "incubating" entry-point may include breaking changes in minor versions, it is recommended that instrumentation libraries not import @opentelemetry/semantic-conventions/incubating in runtime code, but instead copy relevant definitions into their own code base.

hectorhdzg commented on packages/instrumentation-openai/src/internal-types.ts:
```diff
@@ -65,3 +65,111 @@ export type GenAIToolMessageEventBody = {
+ */
+export interface ToolCallResponsePart extends GenericPart {
+  /** The type of the content captured in this part. */
+  type: 'tool_call_response';
```
Looks like you are using this GenericPart "type" in multiple places now, maybe having a type/enum for it would be helpful

btomaj replied:
Sorry, I don't understand what you mean when you say having a type/enum.

For background, I simply converted the [GenAI output JSON schema](https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-output-messages.json) to interfaces. I've tried to be faithful to existing designs/decisions, and restrain my desire to extend without understanding the context of those decisions.

btomaj commented:
> a summary in the PR would be really helpful for reviewers and for future references, something describing the genai.client metrics that are being added, new Spans and Logs and we should summarize this in Readme so users know what kind of telemetry is being created when using this instrumentation

Thank you for doing a review @hectorhdzg. I will make these updates, and come back to you.

hectorhdzg commented:
> > a summary in the PR would be really helpful for reviewers and for future references, something describing the genai.client metrics that are being added, new Spans and Logs
> 
> @hectorhdzg, I have updated the PR with a summary of changes. Please let me know any feedback that you have.
> 
> > we should summarize this in Readme so users know what kind of telemetry is being created when using this instrumentation
> 
> I had a look in the README.md file, and currently, it only says:
> 
> > This package implements Semantic Convention [Version 1.36.0](https://github.com/open-telemetry/semantic-conventions/blob/v1.36.0/docs/README.md).
> 
> I intend to expand this to say:
> 
> > This package implements Semantic Convention [Version 1.36.0](https://github.com/open-telemetry/semantic-conventions/blob/v1.36.0/docs/README.md) for Chat Completions, and Semantic Convention [Version 1.38.0](https://github.com/open-telemetry/semantic-conventions/blob/v1.38.0/docs/README.md) for Responses API.
> 
> Would you like me to go into more detail? If so, given that it will mostly duplicate the information in the Semantic Conventions for Generative AI specification, could you help me judge what level of detail?

@btomaj no need to add more details there for the semantic convention version, people can look at the URL to go deeper if needed, can you update the README of this package at least mentioning what kind of signals will be generated?, I can see Python OpenAI Instrumentation [readme](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai/opentelemetry-instrumentation-openai-v2) is very detailed, but at least talks about that a little bit

btomaj committed:
chore(instrumentation-openai): Adopt 1.38.0 of Semantic Conventions

- Update Semantic Conventions for Generative AI from 1.36.0 to 1.38.0
- Replace hard-coded values with constants from 1.38.0

btomaj committed:
fix(instrumentation-openai): improve types

hectorhdzg commented on packages/instrumentation-openai/src/instrumentation.ts:
```diff
@@ -845,6 +892,938 @@ export class OpenAIInstrumentation extends InstrumentationBase<OpenAIInstrumenta
   }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  _getPatchedResponsesCreate(): any {
```
Take a look at https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3242 to avoid same issue

btomaj replied:
Thanks @hectorhdzg. I'll make sure to update this once I'm unblocked again by the responses to my earlier questions.

hectorhdzg committed:
Merge upstream/main into openai-responses-api

hectorhdzg committed:
Fix Responses API to emit input/output messages as log attributes instead of body, resolve all test TODOs, fix compilation errors, and add system instructions + streaming usage tests.

hectorhdzg committed:
Merge remote-tracking branch 'upstream/main' into openai-responses-api

hectorhdzg committed:
npm install

hectorhdzg committed:
Undo package-lock.json

hectorhdzg committed:
Update

hectorhdzg committed:
Add check for Responses to be available

hectorhdzg committed:
Skip tests where responses not available

hectorhdzg committed:
Lint

hectorhdzg committed:
Merge remote-tracking branch 'upstream/main' into openai-responses-api

hectorhdzg committed:
Improve test coverage

hectorhdzg committed:
Add tests

hectorhdzg committed:
Increase test coverage

hectorhdzg committed:
Lint

JacksonWeber commented on packages/instrumentation-openai/src/instrumentation.ts:
```diff
@@ -845,6 +899,969 @@ export class OpenAIInstrumentation extends InstrumentationBase<OpenAIInstrumenta
+      {
+        [ATTR_GEN_AI_OPERATION_NAME]: GEN_AI_OPERATION_NAME_VALUE_CHAT,
+        [ATTR_GEN_AI_REQUEST_MODEL]: params.model,
+        [ATTR_GEN_AI_PROVIDER_NAME]: GEN_AI_PROVIDER_NAME_VALUE_OPENAI,
```
Makes sense to use the `gen_ai.provider.name` here since sem conv 1.37 deprecated `gen_ai.system`. But this leaves chat/embeddings on the old key while Responses API uses the new attribute. Since I believe these will be shared on the `gen_ai.client.*` metrics, the same instruments might get inconsistent attributes. Might be worth addressing in this PR or flagging for a follow up.

hectorhdzg replied:
Semantic convention updates are outside of the scope of this PR, it cleaner to track somewhere else

trentm commented on packages/instrumentation-openai/test/openai.test.ts:
```
3257:  describeResponses('responses', function () {
3265:    it('adds genai conventions', async () => {
3401:    it('records all the client options', async () => {
3543:    it('records function calls', async () => {
3757:    it('records custom tool calls', async () => {
3911:    it('records file search calls', async () => {
4076:    it('records web search calls', async () => {
4234:    it('records computer use calls', async () => {
4406:    it('records code interpreter calls', async () => {
4560:    it('records image generation call', async () => {
4745:    it('records mcp calls', async () => {
4982:    it('records code interpreter calls with outputs', async () => {
5050:    it('records local shell calls', async () => {
5106:    it('records mcp list tools and approval', async () => {
5174:    it('records system instructions', async () => {
5190:    it('handles connection errors without crashing', async () => {
5273:    it('records function call outputs', async () => {
5356:    it('records multimodal input', async () => {
5399:  describeResponses('responses with content capture', function () {
5407:    it('adds genai conventions', async () => {
5549:    it('records function calls', async () => {
5774:    it('records custom tool calls', async () => {
5934:    it('records file search calls', async () => {
6109:    it('records web search calls', async () => {
6270:    it('records computer use calls', async () => {
6452:    it('records code interpreter calls', async () => {
6614:    it('records image generation call', async () => {
6804:    it('records mcp calls', async () => {
7052:    it('records function call outputs', async () => {
7140:    it('records multimodal input', async () => {
7192:  describeResponses('streaming responses', function () {
7200:    it('adds genai conventions', async () => {
7339:    it('records function calls', async () => {
7596:    it('records custom tool calls', async () => {
7777:    it('records file search calls', async () => {
7976:    it('records web search calls', async () => {
8170:    it('records computer use calls', async () => {
8370:    it('records code interpreter calls', async () => {
8537:    it('records image generation call', async () => {
8740:    it('records mcp calls', async () => {
8990:    it('records usage', async () => {
9035:    it('does not misbehave with double iteration', async () => {
9144:    it('records function call outputs', async () => {
9233:  describeResponses('streaming responses with content capture', function () {
9241:    it('adds genai conventions', async () => {
9386:    it('records function calls', async () => {
9654:    it('records custom tool calls', async () => {
9841:    it('records file search calls', async () => {
10047:    it('records web search calls', async () => {
10246:    it('records computer use calls', async () => {
10451:    it('records code interpreter calls', async () => {
10625:    it('records image generation call', async () => {
10827:    it('records mcp calls', async () => {
11091:    it('does not misbehave with double iteration', async () => {
11206:    it('records function call outputs', async () => {
```

This is *8000* lines of new tests.
Is it necessary to add *all* these?



FWIW, I commented out ~6000 of those lines:

```
% rg '(\bit\b\(|\bdescribe)' test/openai.test.ts
...
3257:  describeResponses('responses', function () {
3265:    it('adds genai conventions', async () => {
3401:    it('records all the client options', async () => {
3543:    it('records function calls', async () => {
3757:    it('records custom tool calls', async () => {
3911:    it('records file search calls', async () => {
4076:    it('records web search calls', async () => {
4234:    it('records computer use calls', async () => {
4406:    it('records code interpreter calls', async () => {
4560:    it('records image generation call', async () => {
4745:    it('records mcp calls', async () => {
4982:    it('records code interpreter calls with outputs', async () => {
5050:    it('records local shell calls', async () => {
5106:    it('records mcp list tools and approval', async () => {
5174:    it('records system instructions', async () => {
5190:    it('handles connection errors without crashing', async () => {
5273:    it('records function call outputs', async () => {
5356:    it('records multimodal input', async () => {
5399:  describeResponses('responses with content capture', function () {
5407:    it('adds genai conventions', async () => {
5549:    // it('records function calls', async () => {
5774:    // it('records custom tool calls', async () => {
5934:    // it('records file search calls', async () => {
6109:    // it('records web search calls', async () => {
6270:    // it('records computer use calls', async () => {
6452:    // it('records code interpreter calls', async () => {
6614:    // it('records image generation call', async () => {
6804:    // it('records mcp calls', async () => {
7052:    // it('records function call outputs', async () => {
7140:    // it('records multimodal input', async () => {
7192:  describeResponses('streaming responses', function () {
7200:    it('adds genai conventions', async () => {
7339:    // it('records function calls', async () => {
7596:    // it('records custom tool calls', async () => {
7777:    // it('records file search calls', async () => {
7976:    // it('records web search calls', async () => {
8170:    // it('records computer use calls', async () => {
8370:    // it('records code interpreter calls', async () => {
8537:    // it('records image generation call', async () => {
8740:    // it('records mcp calls', async () => {
8990:    // it('records usage', async () => {
9035:    // it('does not misbehave with double iteration', async () => {
9144:    // it('records function call outputs', async () => {
9233:  describeResponses('streaming responses with content capture', function () {
9241:    it('adds genai conventions', async () => {
9386:    // it('records function calls', async () => {
9654:    // it('records custom tool calls', async () => {
9841:    // it('records file search calls', async () => {
10047:    // it('records web search calls', async () => {
10246:    // it('records computer use calls', async () => {
10451:    // it('records code interpreter calls', async () => {
10625:    // it('records image generation call', async () => {
10827:    // it('records mcp calls', async () => {
11091:    // it('does not misbehave with double iteration', async () => {
11206:    // it('records function call outputs', async () => {
```

and coverage dropped from `85.18` to `84.95`:

```
  83 passing (4s)
--------------------|---------|----------|---------|---------|--------------------------------------------------------------------------------
File                | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
--------------------|---------|----------|---------|---------|--------------------------------------------------------------------------------
All files           |   88.24 |     66.4 |   81.72 |   88.45 |
 instrumentation.ts |   84.96 |    64.73 |   81.31 |   85.18 | ...,1179,1251-1255,1263-1273,1301-1434,1481-1513,1589-1603,1651-1655,1696,1870
 semconv.ts         |     100 |      100 |     100 |     100 |
 utils.ts           |   95.83 |    93.33 |     100 |   95.83 | 36
--------------------|---------|----------|---------|---------|--------------------------------------------------------------------------------
```

to:

```
  52 passing (4s)
--------------------|---------|----------|---------|---------|--------------------------------------------------------------------------------
File                | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
--------------------|---------|----------|---------|---------|--------------------------------------------------------------------------------
All files           |   88.07 |    58.98 |   81.72 |   88.27 |
 instrumentation.ts |   84.73 |    56.84 |   81.31 |   84.95 | ...,1239,1251-1255,1263-1273,1301-1434,1481-1513,1589-1603,1651-1655,1696,1870
 semconv.ts         |     100 |      100 |     100 |     100 |
 utils.ts           |   95.83 |    93.33 |     100 |   95.83 | 36
--------------------|---------|----------|---------|---------|--------------------------------------------------------------------------------
```

hectorhdzg replied:
Yeah did cleanup here, I was keep getting low coverage [reports](https://app.codecov.io/gh/open-telemetry/opentelemetry-js-contrib/pull/3194?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=open-telemetry), look like there is a different issue causing that

trentm commented on packages/instrumentation-openai/src/instrumentation.ts:
```diff
@@ -734,23 +785,26 @@ export class OpenAIInstrumentation extends InstrumentationBase<OpenAIInstrumenta
     };
   }
 
-  private _getPatchedEmbeddingsCreate(): any {
```
Curious why `private` was removed on this and other methods.
Isn't this going to cause https://github.com/open-telemetry/opentelemetry-js-contrib/pull/3242 again?

trentm commented on packages/instrumentation-openai/src/instrumentation.ts:
```diff
@@ -845,6 +899,969 @@ export class OpenAIInstrumentation extends InstrumentationBase<OpenAIInstrumenta
+  }
+}
+
+class ConvertResponseInputsToInputMessagesUseCase {
```
nit: Could this helper class and the one below be moved to a separate file? Something like responses.ts. Might help for maint.

hectorhdzg committed:
Address comments

trentm commented on packages/instrumentation-openai/test/openai.test.ts:
```diff
@@ -103,6 +112,7 @@ describe('OpenAI', function () {
   });
 
   const client = new OpenAI({ apiKey });
+  const describeResponses = client.responses ? describe : describe.skip;
```
nit: Still needed? Seems like "while I'm still working on this new stuff" dev convenience. Just a nit.

trentm reviewed (approved):
Looks good. I haven't looked carefully at the tests or the specific semver usage compared to spec.

hectorhdzg committed:
Remove describeResponses guard

JacksonWeber commented on packages/instrumentation-openai/src/instrumentation.ts:
```diff
@@ -845,20 +884,304 @@ export class OpenAIInstrumentation extends InstrumentationBase<OpenAIInstrumenta
+          break;
+        }
+        case 'response.output_item.done': {
+          const output = converter.convert([event.item]);
```
Can we wrap the per-event telemetry in a try/catch so the instrumentation can never break the caller's stream?

hectorhdzg committed:
fix: use version check for describeResponses guard

The previous guard (client.responses ? describe : describe.skip) was
dead code since client.responses is always truthy. Use the openai
package VERSION export to properly skip responses tests on versions
prior to v4.87.0 where the Responses API was introduced.

hectorhdzg committed:
Format

hectorhdzg committed:
Add try catch
